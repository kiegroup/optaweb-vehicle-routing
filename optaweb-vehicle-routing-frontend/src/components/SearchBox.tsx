/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { TextInput } from '@patternfly/react-core';
import { OpenStreetMapProvider, SearchResult } from 'leaflet-geosearch';
import * as React from 'react';

interface IProps {
  searchDelay: number;
}

interface IState {
  query: string;
  timeoutId?: number;
}

class SearchBox extends React.Component<IProps, IState> {

  static defaultProps = {
    searchDelay: 500,
  };

  private provider = new OpenStreetMapProvider();

  constructor(props: IProps) {
    super(props);

    this.state = { query: '' };

    this.handleTextInputChange = this.handleTextInputChange.bind(this);
  }

  handleTextInputChange(query: string): void {
    window.clearTimeout(this.state.timeoutId);
    const timeoutId = window.setTimeout(
      async () => {
        const results: SearchResult[] = await this.provider.search({ query });
        console.log(results);
      },
      this.props.searchDelay);
    this.setState({ query, timeoutId });
  }

  render(): React.ReactNode {
    const { query } = this.state;
    return (
      <TextInput
        value={query}
        type="search"
        onChange={this.handleTextInputChange}
        aria-label="geosearch text input"
      />
    );
  }
}

export default SearchBox;
