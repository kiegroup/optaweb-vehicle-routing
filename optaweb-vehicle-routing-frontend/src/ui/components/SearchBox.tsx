/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import '@patternfly/patternfly/patternfly.css';
import { Button, Text, TextContent, TextInput, TextVariants } from '@patternfly/react-core';
import { PlusSquareIcon } from '@patternfly/react-icons';
import { OpenStreetMapProvider } from 'leaflet-geosearch';
import * as React from 'react';
import { LatLng } from 'store/route/types';

export interface Props {
  searchDelay: number;
  boundingBox: [LatLng, LatLng] | null;
  countryCodeSearchFilter: string[];
  addHandler: (result: Result) => void;
}

export interface State {
  query: string;
  results: Result[];
  attributions: string[];
}

export interface Result {
  address: string;
  latLng: LatLng;
}

const searchParams = (props: Props) => ({
  countrycodes: props.countryCodeSearchFilter,
  viewbox: props.boundingBox ?
    [props.boundingBox[0].lng, props.boundingBox[0].lat, props.boundingBox[1].lng, props.boundingBox[1].lat]
    : undefined,
  bounded: !!props.boundingBox,
});

class SearchBox extends React.Component<Props, State> {

  static defaultProps: Pick<Props, 'searchDelay'> = {
    searchDelay: 500,
  };

  private searchProvider: OpenStreetMapProvider;
  private timeoutId: number | null;

  constructor(props: Props) {
    super(props);

    this.state = {
      query: '',
      results: [],
      attributions: [],
    };

    this.searchProvider = new OpenStreetMapProvider({ params: searchParams(props) });
    this.timeoutId = null;

    this.handleTextInputChange = this.handleTextInputChange.bind(this);
    this.handleClick = this.handleClick.bind(this);
  }

  componentWillUpdate(nextProps: Readonly<Props>): void {
    this.searchProvider = new OpenStreetMapProvider({ params: searchParams(nextProps) });
  }

  componentWillUnmount() {
    this.timeoutId && window.clearTimeout(this.timeoutId);
  }

  handleTextInputChange(query: string): void {
    this.timeoutId && window.clearTimeout(this.timeoutId);
    if (query.trim() !== '') {
      this.timeoutId = window.setTimeout(
        async () => {
          const searchResults = await this.searchProvider.search({ query });
          if (this.state.query !== query) {
            return;
          }
          this.setState({
            results: searchResults
              .map(result => ({
                address: result.label,
                latLng: { lat: result.y, lng: result.x },
              })),
            attributions: searchResults
              .map(result => result.raw.licence)
              // filter out duplicate elements
              .filter((value, index, array) => array.indexOf(value) === index),
          });
        },
        this.props.searchDelay,
      );
      this.setState({ query });
    } else {
      this.setState({ query, results: [], attributions: [] });
    }
  }

  handleClick(index: number) {
    this.props.addHandler(this.state.results[index]);
    this.setState({
      query: '',
      results: [],
      attributions: [],
    });
    // TODO focus text input
  }

  render() {
    const { attributions, query, results } = this.state;
    return (
      <>
        <TextInput
          style={{ marginBottom: 10 }}
          value={query}
          type="search"
          placeholder='Search to add a location...'
          aria-label="geosearch text input"
          onChange={this.handleTextInputChange}
        />
        {results.length > 0 && (
          <div className="pf-c-options-menu pf-m-expanded" style={{ zIndex: 1100 }}>
            <ul className="pf-c-options-menu__menu">
              {results.map((result, index) => (
                <li key={`result: ${result}`}>
                  <div className="pf-c-options-menu__menu-item">
                    {result.address}
                    <Button
                      className="pf-c-options-menu__menu-item-icon"
                      variant="link"
                      type="button"
                      onClick={() => this.handleClick(index)}
                    >
                      <PlusSquareIcon />
                    </Button>
                  </div>
                </li>
              ))}

              <li className="pf-c-options-menu__separator" role="separator" />

              {attributions.map((attribution) => (
                <li
                  key={`attrib: ${attribution}`}
                  className="pf-c-options-menu__menu-item pf-m-disabled"
                >
                  <TextContent>
                    <Text
                      component={TextVariants.small}
                    >
                      {attribution}
                    </Text>
                  </TextContent>
                </li>
              ))}
            </ul>
          </div>
        )}
      </>
    );
  }
}

export default SearchBox;
