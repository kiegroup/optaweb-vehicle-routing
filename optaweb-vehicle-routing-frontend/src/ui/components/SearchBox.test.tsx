/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Button, TextInput } from '@patternfly/react-core';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { OpenStreetMapProvider, SearchResult } from 'leaflet-geosearch';
import * as React from 'react';
import SearchBox, { Props, State } from './SearchBox';

jest.mock('leaflet-geosearch');
jest.useFakeTimers();

beforeEach(() => {
  // Clear all instances and calls to constructor and all methods:
  (OpenStreetMapProvider as unknown as jest.MockInstance<OpenStreetMapProvider, []>).mockClear();
});

const searchProviderMock = (
  () => (OpenStreetMapProvider as unknown as jest.MockInstance<OpenStreetMapProvider, []>).mock
);

describe('Search box', () => {
  it('should render text input initially', () => {
    const props: Props = {
      addHandler: jest.fn(),
      boundingBox: null,
      countryCodeSearchFilter: ['XY'],
      searchDelay: 1,
    };
    const searchBox = shallow(<SearchBox {...props} />);
    expect(toJson(searchBox)).toMatchSnapshot();
  });

  it('should show results when query is entered', async () => {
    const props: Props = {
      addHandler: jest.fn(),
      boundingBox: null,
      countryCodeSearchFilter: ['XY'],
      searchDelay: 1,
    };

    const searchBox = shallow(<SearchBox {...props} />);
    expect(searchProviderMock().instances).toHaveLength(1);
    // text input change triggers a component update
    searchBox.find(TextInput).simulate('change', 'London');
    // which in turn creates a new searchProvider instance
    expect(searchProviderMock().instances).toHaveLength(2);
    // so we can't provide the mock implementation earlier than here
    searchProviderMock().instances[1].search = jest.fn().mockImplementation(() => results);
    await jest.runAllTimers();
    expect(searchProviderMock().instances[1].search).toHaveBeenCalledTimes(1);
    expect((searchBox.state() as State).results).toHaveLength(results.length);
    expect((searchBox.state() as State).attributions).toEqual(licenses);
    expect(toJson(searchBox)).toMatchSnapshot();
  });

  it('should hide results when query is empty', () => {
    const props: Props = {
      addHandler: jest.fn(),
      boundingBox: null,
      countryCodeSearchFilter: ['XY'],
      searchDelay: 1,
    };

    const searchBox = shallow(<SearchBox {...props} />);
    expect(searchProviderMock().instances).toHaveLength(1);

    // when there are non-empty results
    searchBox.setState({ results, attributions: licenses });
    expect(searchProviderMock().instances).toHaveLength(2);
    expect((searchBox.state() as State).results).toEqual(results);
    expect((searchBox.state() as State).attributions).toEqual(licenses);

    // and an empty query is issued
    const emptyQuery = ' ';
    searchBox.find(TextInput).simulate('change', emptyQuery);
    expect(searchProviderMock().instances).toHaveLength(3);
    expect(toJson(searchBox)).toMatchSnapshot();

    // search is not invoked
    expect(searchProviderMock().instances[2].search).toHaveBeenCalledTimes(0);
    // and results are cleared
    expect(searchBox.state()).toEqual({ query: emptyQuery, results: [], attributions: [] });
  });

  it('should invoke add handler with the selected result and clear results', () => {
    const mockAddHandler = jest.fn();
    const props: Props = {
      addHandler: mockAddHandler,
      boundingBox: null,
      countryCodeSearchFilter: ['XY'],
      searchDelay: 1,
    };

    const searchBox = shallow(<SearchBox {...props} />);

    // when there are non-empty results
    searchBox.setState({ results, attributions: licenses });
    expect(toJson(searchBox)).toMatchSnapshot();

    const resultItems = searchBox.findWhere(
      node => node.key() !== null && node.key().startsWith('result'),
    );
    expect(resultItems).toHaveLength(results.length);

    const selection = results.length / 2;
    resultItems.at(selection).find(Button).simulate('click');
    expect(props.addHandler).toHaveBeenLastCalledWith(results[selection]);

    expect(searchBox.state()).toEqual({ query: '', results: [], attributions: [] });
  });
});

const licenses = ['License 1', 'License 2'];

const results: SearchResult[] = [{
  label: 'London, ON, Canada',
  x: 101,
  y: 102,
  bounds: [[1, 2], [3, 4]],
  raw: { licence: licenses[0] },
}, {
  label: 'London, OH, USA',
  x: 201,
  y: 202,
  bounds: [[1, 2], [3, 4]],
  raw: { licence: licenses[1] },
}, {
  label: 'London, KY, USA',
  x: 301,
  y: 302,
  bounds: [[1, 2], [3, 4]],
  raw: { licence: licenses[1] },
}, {
  label: 'London, UK',
  x: 401,
  y: 402,
  bounds: [[1, 2], [3, 4]],
  raw: { licence: licenses[0] },
}];
