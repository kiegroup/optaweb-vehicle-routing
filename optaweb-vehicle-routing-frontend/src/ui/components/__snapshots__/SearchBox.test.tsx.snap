// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search box should hide results when query is empty 1`] = `
<Fragment>
  <TextInput
    aria-label="geosearch text input"
    className=""
    data-cy="geosearch-text-input"
    isDisabled={false}
    isReadOnly={false}
    isRequired={false}
    isValid={true}
    onChange={[Function]}
    placeholder="Search to add a location..."
    style={
      Object {
        "marginBottom": 10,
      }
    }
    type="search"
    value=" "
  />
</Fragment>
`;

exports[`Search box should invoke add handler with the selected result and clear results 1`] = `
<Fragment>
  <TextInput
    aria-label="geosearch text input"
    className=""
    data-cy="geosearch-text-input"
    isDisabled={false}
    isReadOnly={false}
    isRequired={false}
    isValid={true}
    onChange={[Function]}
    placeholder="Search to add a location..."
    style={
      Object {
        "marginBottom": 10,
      }
    }
    type="search"
    value=""
  />
  <div
    className="pf-c-options-menu pf-m-expanded"
    style={
      Object {
        "zIndex": 1100,
      }
    }
  >
    <ul
      className="pf-c-options-menu__menu"
    >
      <li
        key="result: [object Object]"
      >
        <div
          className="pf-c-options-menu__menu-item"
        >
          <Button
            className="pf-c-options-menu__menu-item-icon"
            data-cy="geosearch-location-item-button-0"
            onClick={[Function]}
            type="button"
            variant="link"
          >
            <PlusSquareIcon
              color="currentColor"
              size="sm"
              title={null}
            />
          </Button>
        </div>
      </li>
      <li
        key="result: [object Object]"
      >
        <div
          className="pf-c-options-menu__menu-item"
        >
          <Button
            className="pf-c-options-menu__menu-item-icon"
            data-cy="geosearch-location-item-button-1"
            onClick={[Function]}
            type="button"
            variant="link"
          >
            <PlusSquareIcon
              color="currentColor"
              size="sm"
              title={null}
            />
          </Button>
        </div>
      </li>
      <li
        key="result: [object Object]"
      >
        <div
          className="pf-c-options-menu__menu-item"
        >
          <Button
            className="pf-c-options-menu__menu-item-icon"
            data-cy="geosearch-location-item-button-2"
            onClick={[Function]}
            type="button"
            variant="link"
          >
            <PlusSquareIcon
              color="currentColor"
              size="sm"
              title={null}
            />
          </Button>
        </div>
      </li>
      <li
        key="result: [object Object]"
      >
        <div
          className="pf-c-options-menu__menu-item"
        >
          <Button
            className="pf-c-options-menu__menu-item-icon"
            data-cy="geosearch-location-item-button-3"
            onClick={[Function]}
            type="button"
            variant="link"
          >
            <PlusSquareIcon
              color="currentColor"
              size="sm"
              title={null}
            />
          </Button>
        </div>
      </li>
      <li
        className="pf-c-options-menu__separator"
        role="separator"
      />
      <li
        className="pf-c-options-menu__menu-item pf-m-disabled"
        key="attrib: License 1"
      >
        <TextContent>
          <Text
            component="small"
          >
            License 1
          </Text>
        </TextContent>
      </li>
      <li
        className="pf-c-options-menu__menu-item pf-m-disabled"
        key="attrib: License 2"
      >
        <TextContent>
          <Text
            component="small"
          >
            License 2
          </Text>
        </TextContent>
      </li>
    </ul>
  </div>
</Fragment>
`;

exports[`Search box should render text input initially 1`] = `
<Fragment>
  <TextInput
    aria-label="geosearch text input"
    className=""
    data-cy="geosearch-text-input"
    isDisabled={false}
    isReadOnly={false}
    isRequired={false}
    isValid={true}
    onChange={[Function]}
    placeholder="Search to add a location..."
    style={
      Object {
        "marginBottom": 10,
      }
    }
    type="search"
    value=""
  />
</Fragment>
`;

exports[`Search box should show results when query is entered 1`] = `
<Fragment>
  <TextInput
    aria-label="geosearch text input"
    className=""
    data-cy="geosearch-text-input"
    isDisabled={false}
    isReadOnly={false}
    isRequired={false}
    isValid={true}
    onChange={[Function]}
    placeholder="Search to add a location..."
    style={
      Object {
        "marginBottom": 10,
      }
    }
    type="search"
    value="London"
  />
  <div
    className="pf-c-options-menu pf-m-expanded"
    style={
      Object {
        "zIndex": 1100,
      }
    }
  >
    <ul
      className="pf-c-options-menu__menu"
    >
      <li
        key="result: [object Object]"
      >
        <div
          className="pf-c-options-menu__menu-item"
        >
          London, ON, Canada
          <Button
            className="pf-c-options-menu__menu-item-icon"
            data-cy="geosearch-location-item-button-0"
            onClick={[Function]}
            type="button"
            variant="link"
          >
            <PlusSquareIcon
              color="currentColor"
              size="sm"
              title={null}
            />
          </Button>
        </div>
      </li>
      <li
        key="result: [object Object]"
      >
        <div
          className="pf-c-options-menu__menu-item"
        >
          London, OH, USA
          <Button
            className="pf-c-options-menu__menu-item-icon"
            data-cy="geosearch-location-item-button-1"
            onClick={[Function]}
            type="button"
            variant="link"
          >
            <PlusSquareIcon
              color="currentColor"
              size="sm"
              title={null}
            />
          </Button>
        </div>
      </li>
      <li
        key="result: [object Object]"
      >
        <div
          className="pf-c-options-menu__menu-item"
        >
          London, KY, USA
          <Button
            className="pf-c-options-menu__menu-item-icon"
            data-cy="geosearch-location-item-button-2"
            onClick={[Function]}
            type="button"
            variant="link"
          >
            <PlusSquareIcon
              color="currentColor"
              size="sm"
              title={null}
            />
          </Button>
        </div>
      </li>
      <li
        key="result: [object Object]"
      >
        <div
          className="pf-c-options-menu__menu-item"
        >
          London, UK
          <Button
            className="pf-c-options-menu__menu-item-icon"
            data-cy="geosearch-location-item-button-3"
            onClick={[Function]}
            type="button"
            variant="link"
          >
            <PlusSquareIcon
              color="currentColor"
              size="sm"
              title={null}
            />
          </Button>
        </div>
      </li>
      <li
        className="pf-c-options-menu__separator"
        role="separator"
      />
      <li
        className="pf-c-options-menu__menu-item pf-m-disabled"
        key="attrib: License 1"
      >
        <TextContent>
          <Text
            component="small"
          >
            License 1
          </Text>
        </TextContent>
      </li>
      <li
        className="pf-c-options-menu__menu-item pf-m-disabled"
        key="attrib: License 2"
      >
        <TextContent>
          <Text
            component="small"
          >
            License 2
          </Text>
        </TextContent>
      </li>
    </ul>
  </div>
</Fragment>
`;
